### ISSUE TITLE
FedAMP client loss

### ISSUE THREAD

**Issue by wnma3mz** [TsingZ0/PFL-Non-IID#4]
Hi, thanks for your code, it has inspired me a lot. But when I am reading the FedAMP section, is there a mistake in the loss function here?

According to the existing code, the variable `sub` should be a 0 matrix.

https://github.com/TsingZ0/PFL-Non-IID/blob/a478269e52dc25488dc0c83ce5917b229697f881/system/flcore/clients/clientamp.py#L31-L45

I guess, the correct code should be

```python
 for step in range(max_local_steps):
     # self.old_model: the last round model
     params = copy.deepcopy(weight_flatten(self.old_model))

     if self.train_slow: 
         time.sleep(0.1 * np.abs(np.random.rand())) 
     x, y = self.get_next_train_batch() 
     self.optimizer.zero_grad() 
     output = self.model(x) 
     loss = self.loss(output, y) 
  
     params_ = weight_flatten(self.model) 
     sub = params - params_ 
     loss += self.lamda/self.alphaK/2 * torch.dot(sub, sub) 
```

I hope you have time to reply, I appreciate it!

**Comment by TsingZ0** [comment]
Thank you for your attention! There is indeed an error in this line of code.
https://github.com/TsingZ0/PFL-Non-IID/blob/a478269e52dc25488dc0c83ce5917b229697f881/system/flcore/clients/clientamp.py#L40
I have fixed it in the new version according to Equation 6 in the [paper](https://www.aaai.org/AAAI21Papers/AAAI-5802.HuangY.pdf). 
https://github.com/TsingZ0/PFL-Non-IID/blob/9f3dd6c4ac5b7e8a15d07ef015eb8fd4f90beab3/system/flcore/clients/clientamp.py#L40
Please check it!

**Comment by wnma3mz** [comment]
Thank you for your reply, it seems very easy to understand. But the variable `self.client_u` doesn't seem to be updated? Is the copy operation missing.

**Comment by TsingZ0** [comment]
In this platform, for most of the algorithms, the model distributed by the server is updated using `set_parameters()` for further modification on this procedure :-). 
https://github.com/TsingZ0/PFL-Non-IID/blob/c4fddf7ff93aad3e1243a2c8d3cb4a8a40547601/system/flcore/clients/clientamp.py#L53-L55

**Comment by wnma3mz** [comment]
Thank you very much for your answer! It helps me a lot.
